File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 2/큰 수 만들기/큰 수 만들기.cpp
#include <string>
#include <vector>

using namespace std;

string solution(string number, int k) {
    while(k > 0){
        bool flag = false;
        for(int i=0; i<number.length()-1; i++){
            // 오름차순인 곳 찾기
            if(number[i] < number[i+1]){
                number.erase(number.begin()+i);
                flag = true;
                break;
            }
        }
        
        // 못 찾았으면 제일 뒤 지우기
        if(!flag) number.erase(number.begin()+number.length()-1);
        
        k--;
    }
    
    return number;
}

File: 프로그래머스/알고리즘 고득점 Kit/탐욕법(Greedy)/Level 2/큰 수 만들기/큰 수 만들기.py
# stack 이용해서 풀기

def solution(number, k):
    st = [number[0]]
    
    for i in range(1, len(number)):
        if k <= 0: st.append(number[i])
        else:
            # 안에 든 숫자보다 크면 빼고 넣기
            while k>0 and len(st)>0 and number[i]>st[-1]:
                k -= 1
                st.pop(-1)
            st.append(number[i])
            
    # 횟수 아직 남았으면 뒤에서 빼기
    if k > 0:
        st = st[:-k]
        
    return ''.join(st)
